File: Nov2020/24Nov/PalindromeLinkedList.java
/*
Problem Name: 234. Palindrome Linked List
Problem Link: https://leetcode.com/problems/palindrome-linked-list/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if((head == null) || (head.next == null))
            return true;
        ListNode slow = head, fast = head;
        while((fast != null) && (fast.next != null)){
            slow = slow.next;
            fast = fast.next.next;
        }
        slow = reverse(slow);
        fast = head;
        while(slow != null){
            if(slow.val != fast.val)
                return false;
            slow = slow.next;
            fast = fast.next;
        }
        return true;
    }
    static ListNode reverse(ListNode head){
        ListNode prev = null, next = null, cur = head;
        while(cur != null){
            next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        return prev;
    }
}

File: Nov2020/24Nov/PalindromeLinkedList.cpp
/*
Problem name: 234. Palindrome Linked List
Problem link: https://leetcode.com/problems/palindrome-linked-list/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(head==nullptr || head->next==nullptr)    return true;
        stack<int>s;
        ListNode *pt1=head;
        while(pt1->next!=nullptr){
            s.push(pt1->val);
            pt1=pt1->next;
        }
        s.push(pt1->val);
        pt1=head;
        while(pt1->next!=nullptr){
            if(pt1->val!=s.top()){
                return false;
            }
            s.pop();
            pt1=pt1->next;
        }
        return true;
    }
};
